public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -62,26 +64,95 @@ function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ
     } else {
       const duration = parseDuration(defaultPrometheusQueryOverlapWindow);
-      this.overlapWindow = durationToMilliseconds(duration);
+      this.overlapWindowMs = durationToMilliseconds(duration);
     }
+    this.profile();
   }
 
-  cache = new Map<TargetIdent, TargetCache>();
+  private profile() {
+    // Check if PerformanceObserver is supported, and if we have Faro enabled for internal profiling
+    if (typeof PerformanceObserver === 'function' && config.grafanaJavascriptAgent.enabled) {
+      this.perfObeserver = new PerformanceObserver((list: PerformanceObserverEntryList) => {
+        list.getEntries().forEach((entry) => {
+          // eslint-ignore-next-line
+          const entryTypeCast: PerformanceResourceTiming = entry as PerformanceResourceTiming;
+
+          // Safari support for this is coming in 16.4:
+          // https://caniuse.com/mdn-api_performanceresourcetiming_transfersize
+          // Gating that this exists to prevent runtime errors
+          const isSupported = typeof entryTypeCast?.transferSize === 'number';
+
+          if (entryTypeCast?.initiatorType === 'fetch' && isSupported) {
+            let fetchUrl = entryTypeCast.name;
+
+            if (fetchUrl.includes('/api/ds/query')) {
+              let match = fetchUrl.match(/requestId=([a-z\d]+)/i);
+
+              if (match) {
+                let requestId = match[1];
+
+                // TODO: store full initial request size by targSig so we can diff follow-up partial requests
+                // TODO: log savings between full initial request and incremental request to Faro
+                const requestTransferSize = Math.round(entryTypeCast.transferSize);
+                const idToBytes = this.pendingRequestIdsToTargSig[requestId];
+
+                if (idToBytes) {
+                  // Set the transfer size for this current request
+                  Object.keys(idToBytes).forEach((key) => {
+                    for (let otherRequestIds in this.pendingRequestIdsToTargSig) {
+                      const firstKey = Object.keys(this.pendingRequestIdsToTargSig[otherRequestIds])[0];
+                      const firstValue = Object.values(this.pendingRequestIdsToTargSig[otherRequestIds])[0];
+                      if (firstKey === key && firstValue !== null) {
+                        faro.api.pushEvent(
+                          'prometheus incremental query response size',
+                          {
+                            id: firstKey,
+                            initialRequestBytes: firstValue.toString(),
+                            subsequentRequestBytes: requestTransferSize.toString(10),
+                          },
+                          'no-interaction',
+                          {
+                            skipDedupe: true,
+                          }
+                        );
+
+                        delete this.pendingRequestIdsToTargSig[requestId];
```

Comment:
it's better to use a `new Map()` for this and `this.pendingRequestIdsToTargSig.delete(requestId);`